dhis-2/dhis-web-api/src/test/java/org/hisp/dhis/webapi/controller/DatesBindingTest.java

```diff
@@ -0,0 +1,115 @@
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+import java.util.Date;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import org.hisp.dhis.dxf2.webmessage.WebMessage;
+import org.hisp.dhis.util.DateUtils;
+import org.hisp.dhis.webapi.webdomain.EndDateTime;
+import org.hisp.dhis.webapi.webdomain.StartDateTime;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.springframework.stereotype.Controller;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+class DatesBindingTest {
+  private static final String ENDPOINT = "/binding";
+
+  private MockMvc mockMvc;
+  private Date actualStartDateTime;
+  private Date actualEndDateTime;
+
+  @BeforeEach
+  public void setUp() {
+    mockMvc = MockMvcBuilders.standaloneSetup(new BindingController()).build();
+  }
+
+  @Test
+  void shouldReturnADateAtTheEndOfTheDayWhenAnEndDateIsPassedWithoutTime() throws Exception {
+    mockMvc.perform(get(ENDPOINT).param("before", "2001-06-17")).andExpect(status().isOk());
+
+    assertNull(actualStartDateTime);
+    assertEquals("2001-06-17T23:59:59", DateUtils.toLongDate(actualEndDateTime));
+  }
+
+  @Test
+  void shouldReturnADateWithTimeWhenAnEndDateIsPassedWithTime() throws Exception {
+    mockMvc
+        .perform(get(ENDPOINT).param("before", "2001-06-17T16:45:34"))
+        .andExpect(status().isOk());
+
+    assertNull(actualStartDateTime);
+    assertEquals("2001-06-17T16:45:34", DateUtils.toLongDate(actualEndDateTime));
+  }
+
+  @Test
+  void shouldReturnADateAtTheStartOfTheDayWhenAnStartDateIsPassedWithoutTime() throws Exception {
+    mockMvc.perform(get(ENDPOINT).param("after", "2001-06-17")).andExpect(status().isOk());
+
+    assertEquals("2001-06-17T00:00:00", DateUtils.toLongDate(actualStartDateTime));
+    assertNull(actualEndDateTime);
+  }
+
+  @Test
+  void shouldReturnADateWithTimeWhenAnStartDateIsPassedWithTime() throws Exception {
+    mockMvc.perform(get(ENDPOINT).param("after", "2001-06-17T16:45:34")).andExpect(status().isOk());
+
+    assertEquals("2001-06-17T16:45:34", DateUtils.toLongDate(actualStartDateTime));
+    assertNull(actualEndDateTime);
+  }
+
+  @Controller
+  private class BindingController {
+    @GetMapping(value = ENDPOINT)
+    public @ResponseBody WebMessage getDefault(Params params) {
+      actualStartDateTime = applyIfNotNull(params.getAfter(), StartDateTime::toDate);
+      actualEndDateTime = applyIfNotNull(params.getBefore(), EndDateTime::toDate);
+      return ok();
+    }
+  }
+
+  @NoArgsConstructor
+  @Data
+  private class Params {
+    private StartDateTime after;
```

Comment:
`start` and `end` would probably make it easier to read the test as one does not need to translate what `before` refers to.